src/solver/optimize.cpp

```diff
@@ -1,14 +1,294 @@
+#include <math.h>
+#include <string>  
+#include <iostream> 
+#include <sstream> 
 #include "../components/token.hpp"
 #include "../components/order.hpp"
 #include "../components/amm.hpp"
 #include "./optimize.hpp"
+#include "../../external_libs/json.hpp"
 
-void solve_auction(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms)
+
+// auxiliary function (used temporarily) for computing x^n using the boost library data types
+boost::multiprecision::mpf_float custom_pow(boost::multiprecision::mpz_int x, int n)
+{
+    int i;
+    boost::multiprecision::mpf_float res = 1;
+    int m = (n > 0) ? n : (-n);
+
+    for (i = 0; i < m; i++)
+        res = res * x;
+    if (n > 0)
+        return res;
+    else
+        return (1/res);
+}
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
-    // TO BE COMPLETED
-    int *vertices;
-    int **edges;
+   int j, num_amms;
+    boost::multiprecision::mpz_int x0, y0, x1, y1, x_, y_, x;
+    boost::multiprecision::mpf_float rational_x1;
```

Comment:
I am a bit confused by the name, is this mpf_float a float or a rational?

Reply:
Sorry for the confusion. The name just meant that this is a float version of the mpz_int variable x1. Bad choice of name, I agree. I fixed it.